src/controller/CHIPDeviceController.cpp

```diff
@@ -1026,6 +1030,103 @@ CHIP_ERROR DeviceCommissioner::UnpairDevice(NodeId remoteDeviceId)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR DeviceCommissioner::SendEchoRequest(NodeId remoteDeviceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Messaging::SendFlags sendFlags;
+    System::PacketBufferHandle payloadBuf;
+    char requestData[32];
+
+    uint32_t size = mPingParams.GetEchoReqSize();
+    VerifyOrExit(size >= 32 && size <= kMaxPayloadSize, err = CHIP_ERROR_MESSAGE_TOO_LONG);
+
+    snprintf(requestData, 32, "Echo Message %" PRIu32 "\n", mPingParams.GetEchoReqCount() + 1);
+
+    payloadBuf = MessagePacketBuffer::NewWithData(requestData, size);
+    VerifyOrExit(!payloadBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    if (mPingParams.IsUsingMRP())
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNone);
+    }
+    else
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNoAutoRequestAck);
+    }
+
+    mPingParams.SetLastEchoTime(System::Timer::GetCurrentEpoch());
+
+    ChipLogDetail(Controller, "Send echo request message with payload size: %d bytes to Node: %" PRIu64, size, remoteDeviceId);
+
+    err = mEchoClient.SendEchoRequest(std::move(payloadBuf), sendFlags);
+
+    if (err == CHIP_NO_ERROR)
+    {
+        mPingParams.SetWaitingForEchoResp(true);
```

Comment:
How do you know you haven't gotten the response yet at this point?  Seems to me we should set this before we make the SendEchoRequest call and unset it on error, or do some sort of synchronization to ensure message processing can't happen under SendEchoRequest.

Reply:
It rely on mPingParams.Reset(); at the beginning of this function to first set this flag to false.  I agree, Reset this flag to false before we make each SendEchoRequest call is better.

Reply:
No, the point is that you want to set the flag to true _before_ you send the message, as things stand.

Reply:
I am confused, the current logic is to reset the WaitingForEchoResp flag to false before we send the echo request, and change it to true after the echo request has been successfully sent.

Reply:
See:

> How do you know you haven't gotten the response yet at this point?

And the answer, given our current threading situation is that you don't.

Reply:
Acked